mecanum_drive_controller/test/test_mecanum_drive_controller.cpp

```diff
@@ -0,0 +1,547 @@
+
+  ASSERT_TRUE(controller_->wait_for_commands(executor));
+  EXPECT_FALSE(std::isnan((*(controller_->input_ref_.readFromNonRT()))->twist.linear.x));
+  EXPECT_FALSE(std::isnan((*(controller_->input_ref_.readFromNonRT()))->twist.angular.z));
+  EXPECT_EQ((*(controller_->input_ref_.readFromNonRT()))->twist.linear.x, 1.5);
+  EXPECT_EQ((*(controller_->input_ref_.readFromNonRT()))->twist.linear.y, 0.0);
+  EXPECT_EQ((*(controller_->input_ref_.readFromNonRT()))->twist.angular.z, 0.0);
+}
+
+TEST_F(MecanumDriveControllerTest, test_update_logic_chainable)
+{
+  // 1. age>ref_timeout 2. age<ref_timeout
+  SetUpController();
+  rclcpp::executors::MultiThreadedExecutor executor;
+  executor.add_node(controller_->get_node()->get_node_base_interface());
+
+  ASSERT_EQ(controller_->on_configure(rclcpp_lifecycle::State()), NODE_SUCCESS);
+  controller_->set_chained_mode(false);
+  ASSERT_EQ(controller_->on_activate(rclcpp_lifecycle::State()), NODE_SUCCESS);
+  ASSERT_FALSE(controller_->is_in_chained_mode());
+
+  // set command statically
+  static constexpr double TEST_LINEAR_VELOCITY_X = 1.5;
+  static constexpr double TEST_LINEAR_VELOCITY_y = 0.0;
+  static constexpr double TEST_ANGULAR_VELOCITY_Z = 0.0;
+  joint_command_values_[1] = 111;
+  std::shared_ptr<ControllerReferenceMsg> msg = std::make_shared<ControllerReferenceMsg>();
+
+  msg->header.stamp = controller_->get_node()->now() - controller_->ref_timeout_ -
+                      rclcpp::Duration::from_seconds(0.1);
+  msg->twist.linear.x = TEST_LINEAR_VELOCITY_X;
+  msg->twist.linear.y = TEST_LINEAR_VELOCITY_y;
+  msg->twist.linear.z = std::numeric_limits<double>::quiet_NaN();
+  msg->twist.angular.x = std::numeric_limits<double>::quiet_NaN();
+  msg->twist.angular.y = std::numeric_limits<double>::quiet_NaN();
+  msg->twist.angular.z = TEST_ANGULAR_VELOCITY_Z;
+  controller_->input_ref_.writeFromNonRT(msg);
+  const auto age_of_last_command =
+    controller_->get_node()->now() - (*(controller_->input_ref_.readFromNonRT()))->header.stamp;
+
+  // age_of_last_command > ref_timeout_
+  ASSERT_FALSE(age_of_last_command <= controller_->ref_timeout_);
+  ASSERT_EQ((*(controller_->input_ref_.readFromRT()))->twist.linear.x, TEST_LINEAR_VELOCITY_X);
+  ASSERT_EQ(
+    controller_->update_reference_from_subscribers(
+      controller_->get_node()->now(), rclcpp::Duration::from_seconds(0.01)),
+    controller_interface::return_type::OK);
+  ASSERT_EQ(
+    controller_->update_and_write_commands(
+      controller_->get_node()->now(), rclcpp::Duration::from_seconds(0.01)),
+    controller_interface::return_type::OK);
+
+  // EXPECT_TRUE(std::isnan(joint_command_values_[NR_CMD_ITFS - 2]));
+  EXPECT_TRUE(std::isnan(controller_->reference_interfaces_[0]));
+
+  EXPECT_EQ(joint_command_values_[1], 0);
+  // ASSERT_EQ(controller_->reference_interfaces_[0], 0);
+  for (const auto & interface : controller_->reference_interfaces_)
+  {
+    // ASSERT_EQ(interface, 0);
+    EXPECT_TRUE(std::isnan(interface));
```

Comment:
Shouldn't this be 0?

Reply:
reference_interfaces are set to nan and command_interfaces to 0

here we are checking reference_interfaces in loop

Reply:
now added to check command_interfaces =0 if ref_timeout case